sections/03_testing_code.qmd

```diff
@@ -1,3 +1,13 @@
--   Unit tests
--   Integration tests
--   Test-driven development
+When we are conducting experiments we often conduct rigorous testing to ensure that results are reliable and accurate - why should that be any different when you are developing your code? By testing your code you can be sure that it is working as intended, as well as identify any limitations of the code. While the idea of developing tests may feel a bit excessive when starting out with coding it is valuable to be aware of the principles as it still provides a conceptual basis from which you can develop code that meets the expectations associated with conducting 'good' science.
+
+### Unit tests
+
+Unit testing focuses on testing *parts* of code to ensure that a function is doing what it should be doing and meets the specified requirements in a formalised and automated manner. At a high level one can do this by inputting a value into the function for which you know what the output (answer) should be and testing if the answer that the function gives you is the same. Of course this might not always be that easy since you might have many moving parts to a function, in which case the aim should be to 'control' a subset of the parts and test others.
```

Comment:
- In the first sentence: You say that unit tests tests "parts" of the code. I am not sure if this is true, but unit tests always test individual functions right? If yes, maybe it's clearer to write something like "Unit testing focuses on testing *individual functions* of your codebase to ensure that they do what they should be ....." Otherwise the "parts" part of the sentence might be confusing to someone who is used to a data analysis workflow without any functions and they might wonder what those "parts" are
- I am not sure I understand the last sentence. Do you mean writing different test cases for the same function if it has e.g. many arguments?

Reply:
Yeah for the last point I was thinking more in terms of complicated models where you might have multiple parameters/terms and you want to ensure that each individual parameter 'behaves' correctly. But I think the previous information is probably already sufficient